search/executor/iterator.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package executor
+
+import (
+	"errors"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/search"
+	"github.com/m3db/m3ninx/util"
+)
+
+var (
+	errNotEnoughReaders = errors.New("searcher returned more postings lists than the number of readers")
+	errTooManyReaders   = errors.New("searcher returned less postings lists than the number of readers")
+)
+
+type iterator struct {
+	refCount util.RefCount
+	searcher search.Searcher
+	readers  index.Readers
+
+	idx      int
+	currDoc  doc.Document
+	currIter doc.Iterator
+
+	err    error
+	closed bool
+}
+
+func newIterator(rc util.RefCount, s search.Searcher, rs index.Readers) (doc.Iterator, error) {
+	it := &iterator{
+		refCount: rc,
+		searcher: s,
+		readers:  rs,
+	}
+
+	currIter, err := it.nextIter()
+	if err != nil {
+		return nil, err
+	}
+	rc.IncRef()
```

Comment:
looks like this is the only place you incref in this file. Why have the iterator separate from the executor? their lifecycles look 1-1 to me.

Reply:
There's a corresponding `decRef` in the `Close` method. The executor can return multiple iterators which the client will use. Some may be used immediately and closed but others may be used in bursts, for example when processing in batches as we may do when paging so I think it would be beneficial if their lifetimes weren't tied together.